File: generated_code/python/solution_4265_1.py
from typing import List

class Solution:
    def longestBalanced(self, s: str) -> int:
        """
        Intuition:
        A balanced string has an equal number of '0's and '1's. The length of a 
        balanced substring must be 2 * min(count('0'), count('1')) within that 
        substring, because we can always pair up '0's and '1's. 
        
        Since we can perform AT MOST ONE SWAP on the entire string, we can effectively:
        1. Keep the string as is.
        2. Move one '0' from outside our chosen substring to inside it (replacing a '1').
        3. Move one '1' from outside our chosen substring to inside it (replacing a '0').
        4. Swap two elements within the substring (which changes nothing about counts).
        5. Swap an element inside the substring with an identical element outside (changes nothing).
        
        Thus, for any window/substring, the maximum number of '0's we can achieve 
        inside it after at most one swap is:
            max_zeros = zeros_in_window + (1 if total_zeros > zeros_in_window and ones_in_window > 0 else 0)
        Similarly for '1's:
            max_ones = ones_in_window + (1 if total_ones > ones_in_window and zeros_in_window > 0 else 0)
            
        The actual balanceable length for a specific window is 2 * min(max_zeros, max_ones).
        However, the window itself cannot be smaller than this length. In fact, we can just 
        check all possible substrings (or use an optimized sliding window/two-pointer approach) 
        to find the maximum possible balanced length. Given the problem is typical for 
        O(N) or O(N^2), let's optimize using a sliding window or direct check.
        
        Actually, any valid balanced substring of length 2K can be formed if we can get 
        at least K '0's and K '1's in a window of length 2K. 
        Let's find the total counts of '0's and '1's first.
        """
        n = len(s)
        total_0 = s.count('0')
        total_1 = n - total_0
        
        # Max possible balanced length cannot exceed 2 * min(total_0, total_1)
        max_possible_len = 2 * min(total_0, total_1)
        
        # We can binary search the answer (only even lengths) or slide a window.
        # Since we want to maximize the balanced substring, let's check from largest possible even length downwards.
        for length in range(max_possible_len, -1, -2):
            if length == 0:
                return 0
                
            # Count 0s and 1s in the first window of this 'length'
            c0 = s[:length].count('0')
            c1 = length - c0
            
            # Slide the window across the string
            for i in range(n - length + 1):
                if i > 0:
                    # Update counts for the sliding window
                    if s[i - 1] == '0': c0 -= 1
                    else: c1 -= 1
                    if s[i + length - 1] == '0': c0 += 1
                    else: c1 += 1
                
                # Check if we can make this window balanced with at most 1 swap.
                # Case 1: Already balanced
                if c0 == c1:
                    return length
                
                # Case 2: We need one more '0' and we have an available '0' outside, 
                # and we can sacrifice a '1' inside.
                if c0 + 1 == c1 and total_0 > c0:
                    return length
                    
                # Case 3: We need one more '1' and we have an available '1' outside,
                # and we can sacrifice a '0' inside.
                if c1 + 1 == c0 and total_1 > c1:
                    return length
                    
        return 0

File: generated_code/cpp/solution_4265_1.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <unordered_map>

using namespace std;

class Solution {
public:
    int longestBalanced(string s) {
        int n = s.length();
        
        // Count total frequencies of '0' and '1' in the original string
        int cnt0 = 0;
        for (char c : s) {
            if (c == '0') cnt0++;
        }
        int cnt1 = n - cnt0;
        
        // Map to store all indices where each prefix sum occurs.
        // Prefix sum formula: +1 for '1', -1 for '0'.
        unordered_map<int, vector<int>> pos;
        pos[0].push_back(-1); // Base case for prefix sum 0 at index -1
        
        int ans = 0;
        int pre = 0;
        
        for (int i = 0; i < n; ++i) {
            pre += (s[i] == '1') ? 1 : -1;
            pos[pre].push_back(i);
            
            // Case 1: Already balanced without any swap.
            // Substring from pos[pre][0] + 1 to i has an equal number of '0's and '1's.
            ans = max(ans, i - pos[pre][0]);
            
            // Case 2: Substring has two more '1's than '0's (Diff = +2).
            // We can swap one '1' from inside with one '0' from outside to make it balanced.
            if (pos.find(pre - 2) != pos.end()) {
                auto& p = pos[pre - 2];
                // Check if the earliest start index leaves enough '0's outside the substring.
                // The current number of '0's inside this substring is (length - 2) / 2.
                if ((i - p[0] - 2) / 2 < cnt0) {
                    ans = max(ans, i - p[0]);
                } else if (p.size() > 1) {
                    // If the earliest start leaves no '0's outside, try the second earliest start.
                    ans = max(ans, i - p[1]);
                }
            }
            
            // Case 3: Substring has two more '0's than '1's (Diff = -2).
            // We can swap one '0' from inside with one '1' from outside to make it balanced.
            if (pos.find(pre + 2) != pos.end()) {
                auto& p = pos[pre + 2];
                // Check if the earliest start index leaves enough '1's outside the substring.
                // The current number of '1's inside this substring is (length - 2) / 2.
                if ((i - p[0] - 2) / 2 < cnt1) {
                    ans = max(ans, i - p[0]);
                } else if (p.size() > 1) {
                    // If the earliest start leaves no '1's outside, try the second earliest start.
                    ans = max(ans, i - p[1]);
                }
            }
        }
        
        return ans;
    }
};